### PR TITLE
test: test max receipt size using TestLoop

### PR DESCRIPTION
**Pull request by jancionear**
Test the `max_receipt_size` limit added in 8f8a73b6e5184e2e8773c9b6ac641d85e0fd5f8c using TestLoop.

Receipts larger than `4_194_304` bytes should be rejected.

We can't easily generate a 4MB receipt because the transaction size limit is 1.5MB. To deal with this I generate the receipt from within a contract. The contract creates a single receipt with multiple `FunctionCall` actions, each of which has a lot of data in the `args`. All those actions sum up to more than 4MB and the receipt size limit is hit.
The transaction fails properly and the runtime doesn't die when it encounters an invalid receipt.

I had to add the functionality to execute a transaction and wait for the result. I implemented something similar to the old `TestEnv::execute_tx` function.

I also added a "default setup" utility function which generates a basic `TestLoop` setup. This allows to get a test going in 1 line instead of 30.

### CODE REVIEW

File: runtime/near-test-contracts/test-contract-rs/src/lib.rs [hunk 1]
```diff
@@ -1092,11 +1092,7 @@ pub unsafe fn call_yield_resume_read_data_id_from_storage() {
     read_register(0, payload.as_ptr() as u64);
 
     let data_id_key = 42u64.to_le_bytes().to_vec();
-    storage_read(
-        data_id_key.len() as u64,
-        data_id_key.as_ptr() as u64,
-        0
-    );
+    storage_read(data_id_key.len() as u64, data_id_key.as_ptr() as u64, 0);
```
Comment (jancionear) [inline runtime/near-test-contracts/test-contract-rs/src/lib.rs:1095]: This is an unrelated change, `cargo fmt` formatted the file this way.

**Comment by jancionear** [comment]
This PR broke the `test_transaction_limit_for_local_congestion` test. I added more code to `test_contract_rs` and somehow this caused congestion level to drop below the value that was required and [the assert](
https://github.com/near/nearcore/blob/6a6483373554327d941d0731dbd19d2c80ec1a3d/integration-tests/src/tests/client/features/congestion_control.rs#L635) failed:
```
thread 'tests::client::features::congestion_control::test_transaction_limit_for_local_congestion' panicked at integration-tests/src/tests/client/features/congestion_control.rs:634:5:
0.1638847165299328 > 0.5 / 4 failed, V1(CongestionInfoV1 { delayed_receipts_gas: 3277694330598656, buffered_receipts_gas: 0, receipt_bytes: 6176, allowed_shard: 3 })
```
I'm not sure why the congestion level dropped. On `master` `incoming_congestion` is `0.1843`, but on this branch it's only `0.1638`.

I noticed that the size of `test_contract_rs.wasm` increased from `100572` to `101_238`, which is very close to the 100 KiB (102_400) boundary for large receipts, but that should matter only for contract deployment. Maybe it reshuffled something about `allowed_shard`? :thinking: 

I fixed the issue by changing the requirement from `upper_limit_congestion / 3` to ` upper_limit_congestion / 4`. I think the test is still valid with this lower requirement, but I haven't read very deeply into the test.

@jakmeier could you confirm that I can fix the test by lowering the required congestion level?

**Comment by jancionear** [comment]
I also had to adjust the `test_self_delay` test by @nagisa 

It started failing with `--features nightly,test_features`, the recursion depth was 59 instead of 60.
Interestingly it only fails when both features are enabled, with only `nightly` or only `test_features` it passes.

Enabling `test_features` adds more code to the test contract, so I suspect that calling the larger contract is more expensive and the gas limit is reached quicker. `nightly` might change something about gas costs which would make the contract calls use more gas.

I modified the test to accept a range of recursed depths to make the test more resilient to slight changes in gas costs.

**Review by jakmeier — APPROVED** [review]
> I noticed that the size of test_contract_rs.wasm increased from 100572 to 101_238, which is very close to the 100 KiB (102_400) boundary for large receipts, but that should matter only for contract deployment. Maybe it reshuffled something about allowed_shard? 🤔

The test does deploy the contract first, so yeah there can be different timings as to how chunks are filled with receipts.

But yeah this is weird. I didn't expect the test to be so sensitive to the size of the contract and resulting gas cost changes. I can only imagine we hit a rather unlucky case with how it hits forwarding limits for gas congestion and forwarding size limits.

> I fixed the issue by changing the requirement from `upper_limit_congestion / 3` to ` upper_limit_congestion / 4`. I think the test is still valid with this lower requirement, but I haven't read very deeply into the test.
> 
> @jakmeier could you confirm that I can fix the test by lowering the required congestion level?

Yes, that seems to be the best solution. The test is still valid with your changes.

> Enabling test_features adds more code to the test contract, so I suspect that calling the larger contract is more expensive and the gas limit is reached quicker. nightly might change something about gas costs which would make the contract calls use more gas.

Makes sense, yeah. But it still seems a bit on the extreme side. As far I as I know, the additional gas cost for a function call should be only around 1 Mgas per bytes from the `wasm_contract_loading_bytes` parameter. 1 kB would only be around 0.001 Tgas, which doesn't *feel* like it should have this effect, although it certainly could just hit the boundary to reduce the recursion limit by 1. 
I wonder if there are any other gas changes in nightly that makes this more extreme. :shrug: 

But anyway, the PR looks good to me. Someone else should look at the actual new test with the TestLoop framework though.

File: integration-tests/src/test_loop/utils/transactions.rs [hunk 2]
```diff
@@ -201,3 +206,92 @@ pub fn get_node_data<'a>(node_datas: &'a [TestData], account_id: &AccountId) ->
...
+/// For invalid transactions returns an error.
+/// For valid transactions returns the execution result (which could have an execution error inside, check it!).
+pub fn execute_tx(
+    test_loop: &mut TestLoopV2,
+    tx: SignedTransaction,
+    node_datas: &[TestData],
+    maximum_duration: Duration,
+) -> Result<FinalExecutionOutcomeView, InvalidTxError> {
+    // Last node is usually the rpc node
+    let rpc_node_id = node_datas.len().checked_sub(1).unwrap();
```
Comment (shreyan-gupta) [inline integration-tests/src/test_loop/utils/transactions.rs:236]: I was wondering does it matter to which node we send the request to? Why would we want to send to RPC node, and is it true that the last node is the RPC node?

File: integration-tests/src/test_loop/utils/transactions.rs [hunk 3]
```diff
@@ -201,3 +206,92 @@ pub fn get_node_data<'a>(node_datas: &'a [TestData], account_id: &AccountId) ->
...
+
+    let tx_hash = tx.get_hash();
+
+    let process_result = Arc::new(Mutex::new(None));
+    let process_result_clone = process_result.clone();
+
+    node_datas[rpc_node_id].client_sender.clone().send(MessageWithCallback {
```
Comment (shreyan-gupta) [inline integration-tests/src/test_loop/utils/transactions.rs:243]: Checking, do we need to clone client_sender?

File: integration-tests/src/test_loop/utils/transactions.rs [hunk 4]
```diff
@@ -201,3 +206,92 @@ pub fn get_node_data<'a>(node_datas: &'a [TestData], account_id: &AccountId) ->
...
+
+    test_loop.run_until(
+        |tl_data| {
+            let mut processing_done = false;
+            if let Some(processing_outcome) = &*process_result.lock().unwrap() {
+                match processing_outcome.as_ref().unwrap() {
+                    ProcessTxResponse::NoResponse
+                    | ProcessTxResponse::RequestRouted
+                    | ProcessTxResponse::ValidTx => processing_done = true,
+                    ProcessTxResponse::InvalidTx(_err) => {
```
Comment (shreyan-gupta) [inline integration-tests/src/test_loop/utils/transactions.rs:258]: In our test suit, do we ever expect to have invalid transactions sent? Might be a good idea to just panic at the callback site on line 246 instead?

File: integration-tests/src/test_loop/utils/setups.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+use itertools::Itertools;
+use near_chain_configs::test_genesis::TestGenesisBuilder;
+use near_primitives::types::AccountId;
+
+use crate::test_loop::builder::TestLoopBuilder;
+use crate::test_loop::env::TestLoopEnv;
+use crate::test_loop::utils::ONE_NEAR;
+
+/// 2 producers, 2 validators, 1 rpc node, 4 shards, 20 accounts (account{i}) with 10k NEAR each.
+pub fn standard_setup_1() -> TestLoopEnv {
```
Comment (shreyan-gupta) [inline integration-tests/src/test_loop/utils/setups.rs:13]: IMO, it doesn't make sense to over simplify the setup phase. It already feels concise enough to and standardizing it into a function without any parameters probably invites more complexity and layering, and hinders readability.

If you'd really like, we can add more functionality for different types of nodes in the testloop builder instead.

File: integration-tests/src/test_loop/utils/transactions.rs (hunk 4 above)
Comment (shreyan-gupta) [reply integration-tests/src/test_loop/utils/transactions.rs:258]: In either case, we aren't returning the InvalidTxError from this function as suggested by the signature.

File: integration-tests/src/test_loop/utils/transactions.rs (hunk 3 above)
Comment (jancionear) [reply integration-tests/src/test_loop/utils/transactions.rs]: You're right, it works without the `.clone()`. I copied it from other code in `transactions.rs`.
Fixed, removed all such clones in the file.
It's interesting that clippy didn't pick this up :thinking:

File: integration-tests/src/test_loop/utils/transactions.rs (hunk 4 above)
Comment (jancionear) [reply integration-tests/src/test_loop/utils/transactions.rs:258]: In my test I send a transaction that is invalid - one that hits the transaction size limit.
The function does return an `InvalidTxError`, lower in the code. I couldn't do it directly from `run_until` because it returns a boolean.

File: integration-tests/src/test_loop/utils/setups.rs (hunk 5 above)
Comment (jancionear) [reply integration-tests/src/test_loop/utils/setups.rs:13]: 90% of the tests I've written would work just fine on a basic, default blockchain with a few validators and an rpc node. I think it would be useful to have a simple way to create a basic blockchain that most tests can use.
It takes 40 lines of code to set up a testloop test, IMO that isn't concise.
Without a common function all tests will have their own 40 line setup function.

`TestEnv` was easier to set up, just a few lines of code, e.g:
```rust
let num_nodes = 4;
let validators = (0..num_nodes)
    .map(|i| AccountId::try_from(format!("test{}", i + 1)).unwrap())
    .collect::<Vec<_>>();
let mut env = TestEnv::new(vec![validators.clone()], 5, false);
```

Comment (jancionear) [reply integration-tests/src/test_loop/utils/setups.rs:13]: Apart from shorter code, having a standard test setup could make the tests more readable. Reading a test written by someone else is always easier when they use standard, well known conventions, account names, patterns, etc.

File: integration-tests/src/test_loop/utils/transactions.rs (hunk 2 above)
Comment (jancionear) [reply integration-tests/src/test_loop/utils/transactions.rs:236]: > I was wondering does it matter to which node we send the request to? Why would we want to send to RPC node

I wanted to use an RPC node because it tracks all shards. I'm not sure if I can submit a transaction to a node that doesn't track a shard. There's some forwarding logic, but I haven't read into it. Apart from that I need to read the execution outcome, and I think that won't be available on a node that doesn't track the shard.

> and is it true that the last node is the RPC node?

AFAIU the current logic of assigning node roles is that the first few nodes are producers and validators, which track a single shard, ant the rest of them are RPC nodes which track all shards. This means that the last node would be an RPC node.

https://github.com/near/nearcore/blob/9932e95c5c49262189bc4737366ce46bb2b953c8/integration-tests/src/test_loop/builder.rs#L185-L199

(added in https://github.com/near/nearcore/pull/11696 by @wacban)

It's a bit implicit, it would be nicer to have a builder method that specifies which nodes should be the rpc nodes.

Comment (shreyan-gupta) [reply integration-tests/src/test_loop/utils/transactions.rs:236]: Alright! I would have been slightly happier if we had some check in place to confirm whether we are dealing with a node that tracks all shards or not. If this function is used unwisely, it could fail without there being an actual error in logic.

File: integration-tests/src/test_loop/utils/transactions.rs (hunk 4 above)
Comment (shreyan-gupta) [reply integration-tests/src/test_loop/utils/transactions.rs:258]: Aah right, I missed the return statement on line 284

File: integration-tests/src/test_loop/utils/setups.rs (hunk 5 above)
Comment (shreyan-gupta) [reply integration-tests/src/test_loop/utils/setups.rs:13]: Sure, let's go with this but we should make sure to advertise about this function to all new users. I'm still not super convinced but it's definitely out of scope for this PR :)

**Review by shreyan-gupta — APPROVED** [review]

File: integration-tests/src/test_loop/utils/transactions.rs (hunk 2 above)
Comment (jancionear) [reply integration-tests/src/test_loop/utils/transactions.rs:236]: Yeah I think the RPC nodes need some kind of rework. I think ideally you could specify in the builder that some of the clients will be RPCs and then testloop utilities could automatically pick a known RPC node and perform operations on it.
For now I implemented something that works with the existing code.
